src/formatter/terminal_themed.rs

```diff
@@ -0,0 +1,76 @@
+use std::cell::RefCell;
+use std::io::Write;
+use std::sync::Arc;
+use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+use super::{Formatter, HighlightEvent, Theme};
+use crate::constants::HIGHLIGHT_NAMES;
+use crate::Result;
+
+/// A formatter for highlighting into terminal output with ANSI color codes.
+///
+/// - The primary color of each style is used for the text color.
+/// - The secondary color is not used in this formatter.
+///
+/// This formatter holds theme data in an [`Arc<Theme>`], making it cheap to clone and usable
+/// from multiple threads.
+#[derive(Debug)]
+pub struct TerminalFormatter {
+    theme: Arc<Theme>,
+    stdout: RefCell<StandardStream>,
+}
+
+impl TerminalFormatter {
+    /// Create a new [`TerminalFormatter`] with the given theme.
+    pub fn new(theme: Theme) -> Self {
+        Self {
+            theme: Arc::new(theme),
+            stdout: RefCell::new(StandardStream::stdout(ColorChoice::Always)),
+        }
+    }
+
+    fn color_from_hex(&self, hex: &str) -> Color {
+        let rgb = color_from_hex(hex).unwrap_or((255, 255, 255));
+        Color::Rgb(rgb.0, rgb.1, rgb.2)
+    }
+}
+
+impl Formatter for TerminalFormatter {
+    fn write<W>(&self, source: &str, _writer: &mut W, event: HighlightEvent) -> Result<()>
+    where
+        W: std::fmt::Write,
+    {
+        match event {
+            HighlightEvent::Source { start, end } => {
+                let text = &source[start..end];
+                let mut stdout = self.stdout.borrow_mut();
+                stdout.write_all(text.as_bytes())?;
+                stdout.flush()?;
```

Comment:
I would implement `finish` and call `flush` in there - doing it for every event could get mighty expensive (unless the immediacy is important to your use case.)